runtime/compiler/control/HookedByTheJit.cpp

```diff
@@ -6235,6 +6237,62 @@ static int32_t J9THREAD_PROC samplerThreadProc(void * entryarg)
             // compute jit state
             jitStateLogic(jitConfig, compInfo, diffTime); // Update JIT state before going to sleep
 
+            // detect high code cache occupancy and increase initial invocation counts
+            if (!TR::Options::getCountsAreProvidedByUser())
+               {
+               TR::CodeCacheManager *manager = TR::CodeCacheManager::instance();
+               size_t usedCacheSize = manager->getCurrTotalUsedInBytes();
+               size_t threshold = manager->codeCacheConfig().highCodeCacheOccupancyThresholdInBytes();
+               if ((usedCacheSize > threshold) && !highCodeCacheOccupancyThresholdReached)
+                  {
+                  highCodeCacheOccupancyThresholdReached = true;
+                  int32_t count = TR::Options::getHighOccupancyInitialCount();
+                  int32_t aotCount = TR::Options::getAOTHighOccupancyInitialCount();
+                  TR::Options::getCmdLineOptions()->setInitialCount(count);
+                  TR::Options::getCmdLineOptions()->setInitialBCount(count);
+                  TR::Options::getCmdLineOptions()->setInitialMILCount(count);
+                  TR::Options::getAOTCmdLineOptions()->setInitialCount(aotCount);
+                  TR::Options::getAOTCmdLineOptions()->setInitialBCount(aotCount);
+                  TR::Options::getAOTCmdLineOptions()->setInitialMILCount(aotCount);
+                  if (TR::Options::getCmdLineOptions()->isAnyVerboseOptionSet(TR_VerboseJitState, TR_VerbosePerformance))
```

Comment:
Two suggestions:
1. Remove `TR_VerboseJitState` from above, because this logic is not coupled with JITState anymore
2. Reduce the size of the vlog critical section by moving those option settings up (before the lock). This also makes more sense from the vlog message point of view which says that counts have already been modified.

Reply:
I think the option settings are before the lock? Unless I misunderstand you.

Reply:
Yes, something like:
```
                    TR::Options::getCmdLineOptions()->getInitialCount(),
                    TR::Options::getCmdLineOptions()->getInitialBCount(),
                    TR::Options::getCmdLineOptions()->getInitialMILCount(),
                    TR::Options::getAOTCmdLineOptions()->getInitialCount(),
                    TR::Options::getAOTCmdLineOptions()->getInitialBCount(),
                    TR::Options::getAOTCmdLineOptions()->getInitialMILCount());
                    if (TR::Options::getCmdLineOptions()->isAnyVerboseOptionSet(TR_VerbosePerformance))
                       {
                       TR_VerboseLog::CriticalSection lock;
                       TR_VerboseLog::writeLine(TR_Vlog_PERF, "t=%6u JIT entered high code cache occupancy state",
                         (uint32_t)persistentInfo->getStartTime());
                       TR_VerboseLog::writeLine(TR_Vlog_PERF, "JIT changed invocation counts to: JIT: %d %d %d  AOT: %d %d %d",

```

Reply:
Those `getInitialCount()` etc. calls are actually in the `TR_VerboseLog::writeLine` call. I copied what was done in `jitStateLogic`, but I can just replace those calls with what those values are being set to if you like.

Reply:
I misread the code thinking those are `set` methods, but they are getters and should stay with the statement that prints them.
The code looks good as it is.